### PR TITLE
XW-4819 | ABTest of Right Rail recirculation source

### PR DESCRIPTION
**Pull request by JamesHawking**
https://wikia-inc.atlassian.net/browse/XW-4819

### CODE REVIEW

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 1]
```diff
@@ -63,14 +65,20 @@ require([
...
 				modelName: 'wiki',
 				opts: {
-					rule_language : window.wgContentLanguage
+					rule_language: window.wgContentLanguage
 				}
 			}
 		};
 
 	function prepareRailRecirculation(options) {
+		var request;
```
Comment (bkoval) [inline extensions/wikia/Recirculation/js/recirculation.js:74]: New line below

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 2]
```diff
@@ -82,7 +90,72 @@ require([
...
+	}
+
+	function getDataFromRecirculationCMS() {
+		var deferred = $.Deferred();
+
+		nirvana.sendRequest({
+			controller: 'RecirculationApi',
+			method: 'getFandomPosts',
+			format: 'json',
+			type: 'get',
```
Comment (bkoval) [inline extensions/wikia/Recirculation/js/recirculation.js:120]: Not needed?

File: extensions/wikia/Recirculation/services/CurationCMSService.class.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
+<?php
+
+class CurationCMSService {
```
Comment (rybmat) [inline extensions/wikia/Recirculation/services/CurationCMSService.class.php:3]: IMO we should create this class in the same place like for example ArticleVideoService class, `includes/wikia/services` or something like that

File: extensions/wikia/Recirculation/services/CurationCMSService.class.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	static protected $instance = null;
+	private $slug;
+	private $limit;
+
+	const API_BASE = 'https://services.wikia.com/curation-cms/stories/feed/slug/';
```
Comment (rybmat) [inline extensions/wikia/Recirculation/services/CurationCMSService.class.php:8]: if curation-cms is pandora service then we should use swagger to generate service client

File: extensions/wikia/Recirculation/services/CurationCMSService.class.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+	const MCACHE_VER = '1.0';
+	const MCACHE_TIME = 900; // 15 minutes
```
Comment (latata) [inline extensions/wikia/Recirculation/services/CurationCMSService.class.php]: Let's decide about the right cache time.

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 6]
```diff
@@ -82,7 +90,72 @@ require([
...
+				url: post.url
+			});
+		});
+
+		return normalizedData;
+	}
+
+	function isCorrectTestWiki(cityId) {
+		// array of RIGHT_RAIL_RECIRCULATION AB test wikis
+		var testWikisCitiesId = ['2233', '147', '1071836', '3035', '250551', '13346', '410', '1081'];
```
Comment (rybmat) [inline extensions/wikia/Recirculation/js/recirculation.js]: `testCityId`? shorter and more standard

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 7]
```diff
@@ -82,7 +90,72 @@ require([
...
+
+		return testWikisCitiesId.indexOf(cityId) > -1;
+	}
+
+	function getDataFromRecirculationCMS() {
```
Comment (rybmat) [inline extensions/wikia/Recirculation/js/recirculation.js]: `getDataFromCurationCMS`?

File: extensions/wikia/Recirculation/services/CurationCMSService.class.php (hunk 3 above)
Comment (rybmat) [reply extensions/wikia/Recirculation/services/CurationCMSService.class.php:3]: and this controller should only use that service

File: extensions/wikia/Recirculation/services/CurationCMSService.class.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+			$memcKey,
+			self::MCACHE_TIME,
+			function() {
+				return $this->apiRequest();
+			}
+		);
+	}
+
+	private function buildUrl() {
+		return self::API_BASE.$this->slug.'?limit='.$this->limit.'&postType=fandom-article';
```
Comment (latata) [inline extensions/wikia/Recirculation/services/CurationCMSService.class.php]: spaces around `.`

**Review by rybmat — CHANGES_REQUESTED** [review]

**Review by bkoval — APPROVED** [review]

File: extensions/wikia/Recirculation/services/CurationCMSService.class.php (hunk 3 above)
Comment (JamesHawking) [reply extensions/wikia/Recirculation/services/CurationCMSService.class.php:3]: Let's do it after AB test.

File: extensions/wikia/Recirculation/services/CurationCMSService.class.php (hunk 4 above)
Comment (JamesHawking) [reply extensions/wikia/Recirculation/services/CurationCMSService.class.php:8]: Let's do it after AB test.

File: extensions/wikia/Recirculation/js/recirculation.js (hunk 1 above)
Comment (JamesHawking) [reply extensions/wikia/Recirculation/js/recirculation.js:74]: Not anymore 🔨

**Review by rybmat — APPROVED** [review]

File: extensions/wikia/Recirculation/RecirculationApiController.class.php [hunk 9]
```diff
@@ -47,6 +40,20 @@ public function getFandomPosts() {
 		] );
 	}
 
+	private function getDataServiceInstance($type, $cityId) {
```
Comment (xkxd) [inline extensions/wikia/Recirculation/RecirculationApiController.class.php:43]: i'd suggest keeping private methods next to each other, at the bottom of class

File: extensions/wikia/Recirculation/RecirculationApiController.class.php [hunk 10]
```diff
@@ -47,6 +40,20 @@ public function getFandomPosts() {
...
+		if ( $type === 'curated' ) {
+			$dataService = new CuratedContentService();
```
Comment (xkxd) [inline extensions/wikia/Recirculation/RecirculationApiController.class.php:45]: variable is redundant, `return ...;`

File: extensions/wikia/Recirculation/js/recirculation.js (hunk 1 above)
Comment (xkxd) [reply extensions/wikia/Recirculation/js/recirculation.js:74]: i'd suggest keeping not defined variables below the defined ones or rewrite that to
```
var isRecirculationABTest = window.Wikia.AbTest.inGroup('RIGHT_RAIL_RECIRCULATION_SOURCE', 'TOPIC_FEED') &&	getCurationCMSTopic();
var request = isRecirculationABTest ? getDataFromCurationCMS() : liftigniter.prepare(options);
```

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 11]
```diff
@@ -63,26 +65,94 @@ require([
...
+
+		data.posts.forEach(function (post) {
+			normalizedData.items.push({
+				title: post.title,
+				thumbnail: post.thumbnail.url,
+				url: post.url
+			});
+		});
+
+		return normalizedData;
```
Comment (xkxd) [inline extensions/wikia/Recirculation/js/recirculation.js:110]: `.map` will be cleaner
```
function getNormalizedCurationCMSData(data) {
    return data.posts.map(function (post) {
         return {
		title: post.title,
		thumbnail: post.thumbnail.url,
		url: post.url
	};
    });
}
```

File: extensions/wikia/Recirculation/js/recirculation.js [hunk 12]
```diff
@@ -63,26 +65,94 @@ require([
...
+		nirvana.sendRequest({
+			controller: 'RecirculationApi',
+			method: 'getFandomPosts',
+			format: 'json',
+			scriptPath: window.wgCdnApiUrl,
+			data: {
+				type: 'stories',
+				slug: getCurationCMSTopic()
+			},
+			callback: function (data) {
```
Comment (xkxd) [inline extensions/wikia/Recirculation/js/recirculation.js:126]: afair, `sendRequest` returns `$.ajax` which allows you to call `nirvana.sendRequest().done().fail()`

File: extensions/wikia/Recirculation/js/recirculation.js (hunk 2 above)
Comment (JamesHawking) [reply extensions/wikia/Recirculation/js/recirculation.js:120]: Needed.

**Pull request merged**